annotated_error.go

```diff
@@ -0,0 +1,49 @@
+package microerror
+
+import "encoding/json"
+
+type annotatedError struct {
+	error
+
+	Annotation string
+}
+
+func withAnnotation(err error, annotation string) *annotatedError {
+	e, ok := err.(*annotatedError)
+	if ok {
+		e.Annotation = annotation
```

Comment:
> I thought the point of this PR was to have immutable errors

It's only about having `Error` type immutable. The thing is we extract it as a `Cause`. It is important it stays untouched. Otherwise we can't match against it.

I'm not fussed about purity here because it's an internal thing, but I'm fine with returning a new instance though.